Introduce async mojo

bmsantos opened the pull request:
This PR introduces an async mojo for apigen.

Please, DO NOT ACCEPT THIS PR until graphql-java-async is out. At this stage I'm just proposing it for review and further comments and/or improvements.

The necessary graphql-java dependency is still embedded in [async](https://github.com/dminkovsky/graphql-java/tree/async) branch of @dminkovsky fork of graphql-java.

More details on graphql-async available [here](https://github.com/graphql-java/graphql-java/pull/283).

Issue #12

brimworks commented on apigen/src/test/projects/asyncposts/src/test/java/com/disteli/posts/PostsTest.java:
```diff
@@ -0,0 +1,212 @@
+      if (env.getParentType().getName().equals("Post")) {
+        final Post post = env.getSource();
+        if (post.getAuthor().getClass().equals(Author.Unresolved.class)) {
+          return CompletableFuture.completedFuture(authors.values()
```
Why are you iterating over every author to obtain the author with this post id (and your using == instead of equals())?

Seems like this would be simpler, no?

authors.get(post.getAuthor().getId());

...there should also be an NPE check in here.

brimworks commented:
Thanks for the contribution!

I'm still trying to understand how this async stuff is different. Is the only thing different the fact that the DataResolver returns a CompletableFuture<Object>? If so, then wouldn't it make more sense to simply modify the implementation of the ResolverDataFetcher rather than making this fork?

The generated AsyncResolver(s) are not very user-friendly in my opinion.

Thanks,
-Brian

bmsantos commented:
No problem. I really want this maven plugin to be able to do async as well as it will make my life much simpler. 😄 

Maybe there's a way to better integrate all this, and hence, the reason for this discussion. I'm sure you have a better idea on how to best integrate this. I've broken this in two separate Mojo's in order to minimize impact on current code and allow for quick diffs between implementations. We can definitely unify it at any point if that ends-up making more sense. 

The idea with CompletableFuture is that the future has to pass it all the way down from the Resolver to the AsyncExecutionContext. At no point there can be any type that is not wrapped in a future.

I'll be making some updates today. I will be introducing a Vert.x based test case that uses apigen. We can then go from there and use it to further improve and refine the apigen async strategy.

brimworks commented:
Excellent! Thanks for opening up this PR so we can get this discussion going.

My initial thought is that the AsyncResolver interface could be this instead:

```java
import java.util.List;
import java.util.concurrent.CompletableFuture;

public interface AsyncResolver<T> {
    public CompletableFuture<List<T>> resolve(List<T> unresolved);
}
```

Then change `ResolverDataFetcher.java` so the constructor accepts either a `Resolver` or an `AsyncResolver`.

Then change `ResolverDataFetcher.get(DataFetchingEnvironment)` so this last line:

```java
    return replaceResolved(result, resolver.resolve(unresolved).iterator(), depth);
```

behaves the same if a `Resolver` was passed into the constructor, otherwise if an `AsyncResolver` was passed in then it does something like this (not tested!):

```java
    return asyncResolver.resolve(unresolved)
        .thenApplyAsync((resolved) -> replaceResolved(result, resolved.iterator(), depth));
```

We could then add maven configuration for the plugin which chooses to either generate code with AsyncResolver(s) or Resolver(s).

I hope this makes sense. I don't have time at the moment to implement the above, so if you want to take a stab, that would be great.

Thanks,
-Brian

brimworks commented:
Now that I have context (your using this with `vertx-dataloader`), if we implement this as I described in my last comment, then your AsyncResolver implementation would look something like this (untested):

```java
public AuthorResolver implements AsyncResolver<Author> {
    @Inject
    private Vertx vertx;
    @Inject
    private Provider<DataLoader<Author, Author>> dataLoaderProvider;
    public CompletableFuture<List<Author>> resolve(List<Author> unresolved) {
        return VertxCompletableFuture.from(vertx, dataLoaderProvider.get().loadMany(unresolved));
    }
}
```

bmsantos commented:
@brimworks, I've implemented a solution based on your comments and I'm trying to figure out how to have the resolver fetch and return all items found.

For example, this QueryVehicles query for all ```vehicles``` implements:

```java
public class QueryVehiclesImpl implements QueryVehicles {
  @Override
  public List<Vehicle> getVehicles() {
    return Arrays.asList(new Vehicle.Unresolved(-1L));
  }
}
```

And a simple test resolver:

```java
public class TestableVehicleResolver implements Vehicle.AsyncResolver {
  private Map<Long, Vehicle> vehicles;

  public TestableVehicleResolver(Map<Long, Vehicle> vehicles) {
    this.vehicles = vehicles;
  }

  @Override
  public CompletableFuture<List<Vehicle>> resolve(final List<Vehicle> unresolved) {
    return completedFuture(new ArrayList<>(vehicles.values())); // is query all
  }
}
```

The resolver is injected with a list of more than one vehicle. When returned to the ResolverDataFetcher, only the first entry will be included:

```java
...
    public Object replaceResolved(Object result, Iterator<Object> resolved, int depth) {
        if ( depth <= 0 ) {
            return resolved.next();
        }
        List<Object> resolvedResults = new ArrayList<>();
        if ( null == result ) return null;
        for ( Object elm : (List)result ) {  // <---- ONLY ONE ENTRY will be resolved
            resolvedResults.add(replaceResolved(elm, resolved, depth-1));
        }
        return resolvedResults;
    }
...
```

This is happening because the result from the QueryVehiclesImpl seems to have to provide a list with the exact amount of entries to be retrieved by the resolver. In my case, this is just an unknown.

How, using the current implementation, can I indicate to the resolver that I'm interested in retrieving all items and have them all included in the final result?

Thanks,
B.

bmsantos commented:
After some thinking, I've made a simple change to ```replaceResolved()```:

```java
public Object replaceResolved(Object result, Iterator<Object> resolved, int depth) {
        if ( depth <= 0 ) {
            return resolved.next();
        }
        List<Object> resolvedResults = new ArrayList<>();
        if ( null == result ) return null;
        final List list = (List) result;
        if (list.isEmpty()) return newArrayList(resolved); // <-- If result list is empty then use resolved
        for ( Object elm : list ) {
            resolvedResults.add(replaceResolved(elm, resolved, depth-1));
        }
        return resolvedResults;
    }
```

This allowed me to move on using query implementations like:
```java
public class QueryVehiclesImpl implements QueryVehicles {
  @Override
  public List<Vehicle> getVehicles() {
    return emptyList(); // <--- indicates that resolved list is to be used
  }

  @Override
  public Vehicle vehicle(final VehicleArgs args) {
    return new Unresolved(args.getId());
  }
}
```

And the resolver looks like:
```java
public class TestableVehicleResolver implements Vehicle.AsyncResolver {
  private Map<Long, Vehicle> vehicles;

  public TestableVehicleResolver(Map<Long, Vehicle> vehicles) {
    this.vehicles = vehicles;
  }

  @Override
  public CompletableFuture<List<Vehicle>> resolve(final List<Vehicle> unresolved) {

    if (!requireNonNull(unresolved).isEmpty()) {
      return completedFuture(unresolved.stream().map(u -> vehicles.get(u.getId())).collect(toList())); // is argument query by id
    }

    return completedFuture(newArrayList(vehicles.values())); // is query all
  }
}
```

bmsantos commented:
@brimworks, updated code as per your recommendations. When you have some time, please have a look in them. The example app [here](https://github.com/bmsantos/vertx-graphql-example) was also updated to make use of these changes.

Let me know if you find any issue.

brimworks commented on apigen/src/main/java/com/distelli/graphql/apigen/ApiGen.java:
```diff
@@ -31,12 +31,14 @@
         private STGroup stGroup;
         private String guiceModuleName;
         private String defaultPackageName;
+        private String groupFileName = "graphql-apigen.stg";
```
Please revert all changes to this file and incorporate the graphql-apigen-async.stg file into graphql-apigen.stg.

brimworks commented on apigen/src/main/java/com/distelli/graphql/apigen/AsyncApiGenMojo.java:
```diff
@@ -0,0 +1,38 @@
+package com.distelli.graphql.apigen;
```
Please remove the new mojo.

brimworks commented on apigen/src/main/java/com/distelli/graphql/apigen/STModel.java:
```diff
@@ -319,7 +326,7 @@ private DataResolver toDataResolver(Type type) {
             if ( !typeEntry.hasIdField() ) return null;
             DataResolver resolver = new DataResolver();
-            resolver.fieldType = typeName + ".Resolver";
+            resolver.fieldType = typeName + (isAsync ? ".AsyncResolver" : ".Resolver");
```
Can we please change the code so it tries to `@Inject` both the AsyncResolver and non-async Resolver? We can then give priority to the async resolver. Then we can revert all these changes too.

bmsantos commented:
@brimworks, latest commits merge all changes and remove the need for the async mojo. As you've mentioned in your comments, both resolvers are now injected and preference is given to Async one.

Let me know of any issue and/or improvements.

brimworks commented on apigen-deps/src/main/java/com/distelli/graphql/ResolverDataFetcher.java:
```diff
@@ -37,16 +50,19 @@ public ResolverDataFetcher(DataFetcher fetcher, Resolver resolver, int listDepth
         if ( env.getSource() instanceof List ) { // batched.
             result = getBatched(env);
-            if ( null != resolver ) addUnresolved(unresolved, result, ++depth);
+            if (nonNull(resolver) || nonNull(asyncResolver)) addUnresolved(unresolved, result, depth + 1);
```
We still need to increment depth, otherwise replaceResolved() will not work correctly when the field is a List of objects to be resolved.

bmsantos replied:
Correct. Very lame from my end. The tests actually confirm the issue. :(

brimworks commented on apigen-deps/src/main/java/com/distelli/graphql/ResolverDataFetcher.java:
```diff
@@ -37,16 +50,19 @@ public ResolverDataFetcher(DataFetcher fetcher, Resolver resolver, int listDepth
+        return isNull(asyncResolver) ?
+          replaceResolved(result, resolver.resolve(unresolved).iterator(), depth) :
+          asyncResolver.resolve(unresolved)
+            .thenApplyAsync(resolved -> replaceResolved(result, List.class.cast(resolved).iterator(), depth));
```
Just curious if this List.class.cast() is necessary. Since the async resolver is now a CompletableFuture<List<T>>, it seems like that would not be necessary.

brimworks commented on apigen-deps/src/main/java/com/distelli/graphql/ResolverDataFetcher.java:
```diff
@@ -1,23 +1,36 @@
+import graphql.schema.DataFetchingEnvironment;
+import graphql.schema.DataFetchingEnvironmentImpl;
+
+import static com.google.common.collect.Lists.newArrayList;
```
I'm pretty sure that we are only using Guava indirectly through Guice. Given that we have users of this library that do not use Guice but instead use Spring (and thus do not have a dependency on Guice), I think using this library will break those users.

brimworks commented:
Thanks @bmsantos! I think we are getting close here... I assume that I shouldn't merge this in until graphql-java has a released version that supports async resolvers, right?

Thanks,
-Brian

bmsantos commented:
@brimworks, correct. Merging this would make it messy for other users interesting in the non-async implementation.

I'll have look in the latest points you've mentioned. Thanks.

bmsantos commented:
Done. Let me know if there's anything else.

Thanks

bmsantos commented:
@brimworks, I did an initial integration of [Vert.x DataLoader](https://github.com/bmsantos/vertx-graphql-example/blob/master/graphql-service/src/main/java/com/github/bmsantos/graphql/resolvers/VehicleResolver.java) in my code sample.

Everything seems to be working just fine but you will notice that the DataLoader is instantiated on every single resolve. The issue is that, if injected, the DataLoader will cache and accumulate Futures in memory and this is not the intention of the DataLoader cache. Its purpose is to cache only for the duration of the graphql query.

The solution I'm foreseeing for this is by injecting a new instance of the DataLoaders everytime in the context object passed along for the graphql query execution:

```
graphQL.executeAsync(query, operationName, context, variables)
```

Then, have the context object passed to the resolvers so that the appropriate DataLoader is used and cache maintained between resolving requests.

This approach would require changes to apigen in order to propagate the context object to the resolvers.

Let me know what you think. Hopefully you might have a better solution.

Cheers

brimworks commented:
In case anyone else is waiting for this... I'm waiting for graphql-java to do an official release before this gets merged in. Unfortunately, the latest news is that they don't plan to support this until 4.x of graphql-java.

brimworks commented:
@bmsantos this issue has been stagnant for a while, and I know graphql-java is on to much newer releases. Do you still desire this change?

brimworks commented:
Closing, as there has been no response. New pull requests are welcomed :).
